examples/state/custom_transitions.rs

```diff
@@ -0,0 +1,273 @@
+            ..LinearRgba::from(sprite.color)
+        };
+
+        sprite.color = new_color.into();
+    }
+}
+
+// We can restart the game by pressing "R".
+// This will trigger an [`AppState::InGame`] -> [`AppState::InGame`]
+// transition, which will run our custom schedules.
+fn restart_game(input: Res<ButtonInput<KeyCode>>, mut next_state: ResMut<NextState<AppState>>) {
+    if input.just_pressed(KeyCode::KeyR) {
+        // Although we are already in this state setting it again will generate an identity transition.
+        // While default schedules ignore those kinds of transitions, out custom schedules will react to them.
+        next_state.set(AppState::InGame);
+    }
+}
+
+/// This module provides the custom `OnReenter` and `OnReexit` transitions for easy installation.
+mod custom_transitions {
+    use crate::*;
+
+    /// Custom schedule that will behave like `OnEnter`,
+    /// but run even during identity transitions.
+    #[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+    pub struct OnReenter<S: States>(pub S);
+
+    /// Schedule runner which checks conditions and if they're right
+    /// runs out custom schedule.
+    fn run_reenter<S: States>(transition: In<Option<StateTransitionEvent<S>>>, world: &mut World) {
+        // We return early if no transition event happened.
+        let Some(transition) = transition.0 else {
+            return;
+        };
+
+        // If we wanted to ignore identity transitions,
+        // we'd compare `exited` and `entered` here,
+        // and return if they were the same.
+
+        // We check if we actually entered a state.
+        // A [`None`] would indicate that the state was removed from the world.
+        // This only happens in the case of [`SubStates`] and [`ComputedStates`].
+        let Some(entered) = transition.entered else {
+            return;
+        };
+
+        // If all conditions are valid, we run our custom schedule.
+        let _ = world.try_run_schedule(OnReenter(entered));
+    }
+
+    /// Custom schedule that will behave like `OnExit`,
+    /// but run even during identity transitions.
+    #[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+    pub struct OnReexit<S: States>(pub S);
+
+    fn run_reexit<S: States>(transition: In<Option<StateTransitionEvent<S>>>, world: &mut World) {
+        let Some(transition) = transition.0 else {
+            return;
+        };
+        let Some(exited) = transition.exited else {
+            return;
+        };
+
+        let _ = world.try_run_schedule(OnReexit(exited));
+    }
+
+    /// The plugin registers the transitions for one specific state.
+    /// If you use this for multiple states consider:
+    /// - installing the plugin multiple times,
+    /// - create an [`App`] extension method that inserts
+    ///   those transitions during state installation.
+    #[derive(Default)]
+    pub struct IdentityTransitionsPlugin<S: States>(PhantomData<S>);
```

Comment:
I think the flow will be clearer with this definition at the start of this submodule.